src/main/java/be/syntra/devshop/DevshopFront/controllers/ProductController.java

```diff
@@ -103,12 +104,35 @@ public String addSelectedProductFromDetailToCart(@PathVariable Long id) {
         return "redirect:/products/details/" + id;
     }
 
+    @PostMapping("/details/{id}/addreview")
+    public String addReviewToProduct(@PathVariable Long id, @ModelAttribute("review") Review review) {
+        reviewService.submitReview(id, review);
+        return "redirect:/products/details/" + id;
+    }
+
+    @PostMapping("/details/{id}/updatereview")
+    public String updateReviewForProduct(@PathVariable Long id, @ModelAttribute("review") Review review) {
+        reviewService.updateReview(id, review);
+        return "redirect:/products/details/" + id;
+    }
+
+    @PostMapping("/details/{id}/deletereview")
+    public String removeReviewFromProduct(@PathVariable Long id, @ModelAttribute("review") Review review) {
+        reviewService.removeReview(id, review);
+        return "redirect:/products/details/" + id;
+    }
+
     private String getProductDetails(Long id, Model model, Principal user) {
         Product product = productService.findById(id);
         StarRatingDto rating = starRatingService.findByUserNameAndId(id, nullSafe(user));
-        model.addAttribute("rating", rating);
+        if (null != user) {
```

Comment:
extract stream to servicemethod, or even create a rest-call

Reply:
made findByUserNameAndId(productId, userName) in reviewService + wrote test

Reply:
since you did that, you could also use the nullSafe(user) method that's in the ProductController class already

Reply:
not going to change, as discussed